text/3668-async-closure.md

```diff
@@ -0,0 +1,541 @@
+    fn async_call_once(self, args: Args) -> Self::CallOnceFuture;
+}
+
+/// An async-aware version of the [`FnMut`](crate::ops::FnMut) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFnMut<Args>: AsyncFnOnce<Args> {
+    /// Future returned by [`AsyncFnMut::async_call_mut`] and [`AsyncFn::async_call`].
+    type CallRefFuture<'a>: Future<Output = Self::Output>
+    where
+        Self: 'a;
+
+    /// Call the [`AsyncFnMut`], returning a future which may borrow from the called closure.
+    fn async_call_mut(&mut self, args: Args) -> Self::CallRefFuture<'_>;
+}
+
+/// An async-aware version of the [`Fn`](crate::ops::Fn) trait.
+///
+/// All `async fn` and functions returning futures implement this trait.
+pub trait AsyncFn<Args>: AsyncFnMut<Args> {
+    /// Call the [`AsyncFn`], returning a future which may borrow from the called closure.
+    fn async_call(&self, args: Args) -> Self::CallRefFuture<'_>;
+}
+```
+
+### Associated types of `AsyncFn*` traits are not nameable
+
+Note that, unlike what is true today with the current `Fn*` traits, this RFC reserves as an implementation detail the associates types of the `AsyncFn*` traits, and these will not be nameable as part of the stable interface specified by this RFC.
+
+### `async` bound modifier on `Fn()` trait bounds
+
+The `AsyncFn*` traits specified above are nameable via a new `async` bound modifier that is allowed on `Fn` trait bounds. That is, `async Fn*() -> T` desugars to `AsyncFn*() -> T` in bounds, where `Fn*` is one of the three flavors of existing function traits: `Fn`/`FnMut`/`FnOnce`.
+
+This RFC specifies the modification to the _TraitBound_ nonterminal in the grammar:
+
+> **<sup>Syntax</sup>**
+> _TraitBound_ :
+> &nbsp;&nbsp;&nbsp;&nbsp;`?`<sup>?</sup> _ForLifetimes_<sup>?</sup> `async`<sup>?</sup> _TypePath_\
+> &nbsp;&nbsp;| `(` `?`<sup>?</sup> _ForLifetimes_<sup>?</sup> `async`<sup>?</sup> _TypePath_ `)`
+
+Since the grammar doesn't distinguish parenthesized and angle-bracketed generics in `_TypePath_`, `async` as a trait bound modifier will be **accepted** in all trait bounds at _parsing_ time, but it will be **denied** by the compiler _post-expansion_ if it's not attached to a parenthesized `Fn()` trait bound.
+
+Users are able to write `async Fn*() -> T` trait bounds in all positions that trait bounds are allowed, for example:
+
+```rust
+fn test(f: F) where F: async Fn() -> i32 {}
+
+fn apit(f: impl async Fn() -> i32) {}
+
+trait Tr {
+    type Callable: async Fn() -> i32;
+}
+
+// Allowed syntactically; not currently object-safe:
+let _: Box<dyn async Fn()> = todo!();
+```
+
+### When is `async Fn*()` implemented?
+
+All currently-stable callable types (i.e., closures, function items, function pointers, and `dyn Fn*` trait objects) automatically implement `async Fn*() -> T` if they implement `Fn*() -> Fut` for some output type `Fut`, and `Fut` implements `Future<Output = T>`.
```

Comment:
Can you elaborate in this RFC why this rule was chosen, and how this is implemented? When looking at the implementation of for example [`AsyncFnOnce`](https://doc.rust-lang.org/std/ops/trait.AsyncFnOnce.html) on nightly today - there is nothing in the implementation that gives the impression this should work. This seems like it's using some kind of lang built-in to work the way it does?

Specifically, I was looking for an equivalent to `impl AsyncFnOnce for F where F: FnOnce() -> impl Future` on the implementation page, but there is none. Given this section is calling `async Fn*()` a trait, I'm assuming this is about the actual `AsyncFn{,Mut,Once}` traits and not just the `async Fn()` syntax.

Reply:
Can you elaborate in this RFC why this rule was chosen, and how this is implemented?

Sure.

The reason is that we want to be able to pass `async fn` function items and regular non-`async` closures to functions that have `F: async Fn()` trait bounds, otherwise there'll be a schism between APIs that *only* take async closures that doesn't exist in the regular `Fn` trait bound world.

It's implemented via a set of built-in trait bounds. A blanket impl is incoherent since async closures provide their own overlapping `AsyncFn*` implementations.

>  Given this section is calling `async Fn*()` a trait, I'm assuming this is about the actual `AsyncFn{,Mut,Once}` traits and not just the `async Fn()` syntax.

Yep. I use them interchangeably throughout the RFC, since users interact with `async Fn()` trait bounds and not the underlying traits, which are likely to disappear eventually.

Reply:
I've added a commit motivating why `AsyncFn*` are implemented using built-in implementations, and some caveats about the implementation in particular.